hibernate-reactive-core/src/main/java/org/hibernate/reactive/util/impl/CompletionStages.java

```diff
@@ -155,25 +152,25 @@ public static void logSqlException(Throwable t, Supplier<String> message, String
 	 * }
 	 * </pre>
 	 */
-	public static <T> CompletionStage<Void> loop(Iterator<T> iterator, BiFunction<T,Integer,CompletionStage<?>> consumer) {
+	public static <T> CompletionStage<Void> loop(Iterator<T> iterator, BiFunction<T, Integer, CompletionStage<?>> consumer) {
 		if ( iterator.hasNext() ) {
-			AtomicInteger index = new AtomicInteger( 0 );
-			return AsyncTrampoline.asyncWhile( () -> consumer.apply( iterator.next(), index.getAndIncrement() )
-					.thenApply( r -> iterator.hasNext() ));
-		}
-		else {
-			return voidFuture();
+			return AsyncTrampoline.asyncWhile(
+					index -> iterator.hasNext(),
+					index -> consumer
+							.apply( iterator.next(), index )
+							.thenApply( u -> index + 1 ),
+					0
+			).thenAccept( CompletionStages::voidFuture );
```

Comment:
OK, so I like getting rid of the `AtomicInteger`, for sure.

But if you're expecting this to be a performance optimization, I doubt that's true. Because here we're boxing an integer for each step of the loop, which creates garbage.

I would rather see an implementation that avoided boxing loop indices, which I believe is conceptually possible, but we would have to replace `BiFunction` with a different SAM.

Reply:
Well, the goal was to remove the various castings, I guess now we can see if there are some changes compared to before

Reply:
> we would have to replace `BiFunction` with a different SAM.

I'd like to try that if you can give me some more details about it

Reply:
What's a SAM?

Reply:
Although, I did this change because I prefer this approach then having an accumulator around.

Reply:
Sorry, SMI, i've no idea what a SAM is either. :-)

Reply:
That doesn't really help :)

Reply:
Single Method Interface.

i.e. an  `IntBiFunction` interface that accepts `int` instead of `Integer`.

Reply:
Ah, ok, thanks